gh-123401: Fix http.cookies module to support obsolete RFC 850 date format

Konano opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->

<!-- gh-issue-number: gh-123401 -->
* Issue: gh-123401
<!-- /gh-issue-number -->

Konano committed:
pythongh-123401: Fix http.cookies module to support obsolete RFC 850 date format

Konano committed:
Add NEWS entry for http.cookies RFC 850 date format support

Konano committed:
Update NEWS entry to use double backticks for inline literals

Wulian233 commented on Lib/http/cookies.py:
```diff
@@ -426,7 +426,9 @@ def OutputString(self, attrs=None):
     (?P<val>                         # Start of group 'val'
     "(?:[^\\"]|\\.)*"                  # Any doublequoted string
```
```suggestion
    "(?:[^\\"]|\\.)*"                  # Any double-quoted string
```

Konano committed:
style(http/cookies): correct typo in regex comment

Fix a minor typo in the regex comment by adding a hyphen to "double-quoted" for consistency and clarity. No functional changes to the code logic.

Co-authored-by: Wulian <1055917385@qq.com>

picnixz commented on Misc/NEWS.d/next/Library/2024-08-27-18-58-01.gh-issue-123401.t4-FpI.rst:
```diff
@@ -0,0 +1,2 @@
+The ``http.cookies`` module now supports parsing obsolete RFC 850 date
+formats, in accordance with RFC 9110 requirements.
```
```suggestion
The :mod:`http.cookies` module now supports parsing obsolete :rfc:`850`
date formats, in accordance with :rfc:`9110` requirements.
```

picnixz commented on Lib/test/test_http_cookies.py:
```diff
@@ -59,6 +59,39 @@ def test_basic(self):
 
+    def test_obsolete_rfc850_date_format(self):
+        # Test cases with different days and dates in obsolete RFC 850 format
+        test_cases = [
```
Can you include test vectors / examples from RFC 850 if any?

Konano replied:
See https://github.com/python/cpython/pull/123405/commits/592da513efcf76740b033544f4a1e959ce4c2728 
Thanks!

picnixz replied:
This is really nitpicking, but I'd like you to also indicate something like:

```
[
 	# test vectors from RFC 850:
	...
	# some other test cases:
	...
]
```

This would help us detecting whether it's something from the RFC that is failing or not. If you can even include where you found it in the RFC, it would be easier for future debugging if any.

Konano committed:
Update Misc/NEWS.d/next/Library/2024-08-27-18-58-01.gh-issue-123401.t4-FpI.rst

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

Konano committed:
Include two examples from RFC 850

picnixz commented on Misc/NEWS.d/next/Library/2024-08-27-18-58-01.gh-issue-123401.t4-FpI.rst:
```diff
@@ -0,0 +1,2 @@
+The :mod:`http.cookies` module now supports parsing obsolete :rfc:`850`
+date formats, in accordance with :rfc:`9110` requirements.
```
You can credit yourself if you want. Add something like `Patch by YOUR_NAME.` on a separate line.

Konano replied:
```suggestion
date formats, in accordance with :rfc:`9110` requirements.
Patch by Nano Zheng.
```

Konano replied:
Am I doing this correctly?

picnixz replied:
Yes (actually, you can just look at other files and copy how they do it).

picnixz reviewed (commented):
At first glance it seems correct. I'm a bit tired for more in-depth review (and even if I were to approve it, a core dev is needed for merging it) so I'll leave it at it for today.

picnixz commented on Lib/test/test_http_cookies.py:
```diff
@@ -59,6 +59,47 @@ def test_basic(self):
+        # Test cases with different days and dates in obsolete RFC 850 format
+        test_cases = [
+            {
+                'data': 'key=value; expires=Saturday, 01-Jan-83 00:00:00 EST',
```
Ah I think I know why the tests are failing. The examples seem to use EST but the regex seems to expect GMT only. Should we allow more timezones?

Konano replied:
https://www.rfc-editor.org/rfc/rfc9110.html#section-5.6.7-12

Maybe GMT only is right. I will fix the test cases.

Konano committed:
Add comments for rfc850 test cases

Konano committed:
change EST to GMT in test cases

Konano committed:
Update Misc/NEWS.d/next/Library/2024-08-27-18-58-01.gh-issue-123401.t4-FpI.rst

vstinner reviewed (approved):
LGTM. Thanks for the tests.

vstinner committed:
Merge branch 'main' into fix-issue-123401
